transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,874 @@
+
+        let sock_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or_else(|| TransportError::MultiaddrNotSupported(addr.clone()))?;
+        if sock_addr.port() == 0 || sock_addr.ip().is_unspecified() {
+            return Err(TransportError::MultiaddrNotSupported(addr));
+        }
+        let remote = addr.clone(); // used for logging
+        trace!("dialing address: {:?}", remote);
+
+        let udp_mux = if let Some(l) = self.listeners.iter_mut().next() {
+            l.udp_mux.clone()
+        } else {
+            return Err(TransportError::Other(Error::NoListeners));
+        };
+        let se = build_setting_engine(udp_mux, &sock_addr, &our_fingerprint);
+        let api = APIBuilder::new().with_setting_engine(se).build();
+
+        // [`Transport::dial`] should do no work unless the returned [`Future`] is polled. Thus
+        // do the `set_remote_description` call within the [`Future`].
+        Ok(async move {
+            let peer_connection = api
+                .new_peer_connection(config)
+                .map_err(Error::WebRTC)
+                .await?;
+
+            let offer = peer_connection
+                .create_offer(None)
+                .map_err(Error::WebRTC)
+                .await?;
+            debug!("OFFER: {:?}", offer.sdp);
+            peer_connection
+                .set_local_description(offer)
+                .map_err(Error::WebRTC)
+                .await?;
+
+            // Set the remote description to the predefined SDP.
+            let remote_fingerprint = match fingerprint_from_addr(&addr) {
+                Some(f) => fingerprint_to_string(&f),
+                None => return Err(Error::InvalidMultiaddr(addr.clone())),
+            };
+            let server_session_description = render_description(
+                sdp::SERVER_SESSION_DESCRIPTION,
+                sock_addr,
+                &remote_fingerprint,
+                &remote_fingerprint.to_owned().replace(':', ""),
+            );
+            debug!("ANSWER: {:?}", server_session_description);
+            let sdp = RTCSessionDescription::answer(server_session_description).unwrap();
+            // Set the local description and start UDP listeners
+            // Note: this will start the gathering of ICE candidates
+            peer_connection
+                .set_remote_description(sdp)
+                .map_err(Error::WebRTC)
+                .await?;
+
+            // Open a data channel to do Noise on top and verify the remote.
+            let data_channel = peer_connection
+                .create_data_channel(
+                    "data",
+                    Some(RTCDataChannelInit {
+                        id: Some(1), // id MUST match one used during upgrade
+                        ..RTCDataChannelInit::default()
+                    }),
+                )
+                .await?;
+
+            let (tx, mut rx) = oneshot::channel::<Arc<DetachedDataChannel>>();
```

Comment:
These kind of low-level details create a lot of noise when reading the code, which makes it really hard to follow as part of the much more high-level "create connection" use-case.

Can we wrap the `RTCPeerConnection` struct in something that provides an in-between interface for us to use in here? It is unfortunate that the `webrtc` interface is so clunky but I think it would really pay off if we tried to abstract this away.

For example, from what I understand, we basically do the following:

1. sdp handshake
2. create dedicated noise data channel
3. noise handshake

It would be great if the code could reflect this. Something like:

```rust
// Contains the sdp handshake etc
let connection = WebRtcConnection::connect_to(&multiaddr).await?;

// Special case for upgrade with ID 1, could reuse a lower-level API for new data channels in `Connection`.
let data_channel = connection.create_initial_upgrade_data_channel().await?; 

// Fails on bad peer ID etc
perform_noise_handshake(data_channel).await?;
```